packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SubjectGroupViewer/SubjectGroupViewerContainer.js

```diff
@@ -0,0 +1,301 @@
+
+import SVGContext from '@plugins/drawingTools/shared/SVGContext'
+import SubjectGroupViewer from './SubjectGroupViewer'
+import locationValidator from '../../helpers/locationValidator'
+import withKeyZoom from '../../../withKeyZoom'
+
+function preventDefault (e) {
+  e.preventDefault()
+}
+
+const DEFAULT_CELL_WIDTH = 1200
+const DEFAULT_CELL_HEIGHT = 1000
+const DEFAULT_CELL_STYLE = {
+  stroke: '#fff',
+  strokeWidth: '4',
+  fill: '#000'
+}
+const DEFAULT_GRID_COLUMNS = 3
+const DEFAULT_GRID_ROWS = 3
+
+function storeMapper (stores) {
+  const {
+    setOnZoom,
+    setOnPan
+  } = stores.classifierStore.subjectViewer
+  
+  const {
+    active: activeWorkflow
+  } = stores.classifierStore.workflows
+  
+  const workflowConfig = activeWorkflow && activeWorkflow.configuration || {}
+  
+  const cellWidth = workflowConfig.cell_width || DEFAULT_CELL_WIDTH
+  const cellHeight = workflowConfig.cell_height || DEFAULT_CELL_HEIGHT
+  const cellStyle = workflowConfig.cell_style || DEFAULT_CELL_STYLE
+  const gridColumns = workflowConfig.grid_columns || DEFAULT_GRID_COLUMNS
+  const gridRows = workflowConfig.grid_rows || DEFAULT_GRID_ROWS
+
+  return {
+    cellWidth,
+    cellHeight,
+    cellStyle,
+    gridColumns,
+    gridRows,
+    setOnZoom,
+    setOnPan
+  }
+}
+
+class SubjectGroupViewerContainer extends React.Component {
+  constructor () {
+    super()
+    this.dragMove = this.dragMove.bind(this)
+    this.groupViewer = React.createRef()
+    this.scrollContainer = React.createRef()
+
+    this.state = {
+      images: [],
+      panX: 0,
+      panY: 0,
+      zoom: 1,
+    }
+  }
+
+  componentDidMount () {
+    this.onLoad()
+
+    // Listen for pan and zoom actions outside of this component.
+    // i.e. zoom in/out actions from the iamge controls. 
+    this.props.setOnPan(this.onPanViaExternalControls.bind(this))
+    this.props.setOnZoom(this.onZoomViaExternalControls.bind(this))
+    
+    this.scrollContainer.current
+    && this.scrollContainer.current.addEventListener('wheel', preventDefault)
+  }
+  
+  componentWillUmount () {
+
+    this.setOnPan(() => true)
+    this.setOnZoom(() => true)
+    
+    this.scrollContainer.current
+    && this.scrollContainer.current.removeEventListener('wheel', preventDefault)
+  }
+
+  fetchImage (url) {
+    const { ImageObject } = this.props
+    return new Promise((resolve, reject) => {
+      const img = new ImageObject()
+      img.onload = () => resolve(img)
+      img.onerror = reject
+      img.src = url
+      return img
+    })
+  }
+
+  dragMove (event, difference) {
+    this.onDrag && this.onDrag(event, difference)
+  }
+
+  async preload () {
+    const { subject } = this.props
+    if (subject && subject.locations) {
+      // TODO: Validate for allowed image media mime types
+      
+      const imageUrls = subject.locations.map(obj => Object.values(obj)[0])
+      const images = await Promise.all(
+        imageUrls.map(url => this.fetchImage(url))
+      )
```

Comment:
Oh wait, `onLoad` wraps this in a `try… catch` so that should be fine.

Reply:
Yup, I was basing the error-handling on the SingleImageViewerContainer. In case either `preload()` or `fetchImage()` throws an error, we'll get the general purpose `loadingState === asyncStates.error` state and the appropriate "Something Went Wrong" error message.

⚠️ Note: this means if ANY of the Subject's sub-Subject's images don't load properly, the WHOLE grid considers it a failure. I thought this made sense, since if I allowed a "just show everything that loaded properly" approach, it might accidentally introduce an untraceable bias when 1 or more images of the N total images was a 404 while the others displayed perfectly.